src/worker.cpp

```diff
@@ -1169,24 +1169,44 @@ types::VPDMapVariant Worker::parseVpdFile(const std::string& i_vpdFilePath)
 
     std::shared_ptr<Parser> vpdParser = std::make_shared<Parser>(i_vpdFilePath,
                                                                  m_parsedJson);
-    types::VPDMapVariant l_parsedVpd = vpdParser->parse();
-
-    // Before returning, as collection is over, check if FRU qualifies for
-    // any post action in the flow of collection.
-    // Note: Don't change the order, post action needs to be processed only
-    // after collection for FRU is successfully done.
-    if (jsonUtility::isActionRequired(m_parsedJson, i_vpdFilePath, "postAction",
-                                      "collection"))
+    try
     {
-        if (!processPostAction(i_vpdFilePath, "collection", l_parsedVpd))
+        types::VPDMapVariant l_parsedVpd = vpdParser->parse();
+
+        // Before returning, as collection is over, check if FRU qualifies for
+        // any post action in the flow of collection.
+        // Note: Don't change the order, post action needs to be processed only
+        // after collection for FRU is successfully done.
+        if (jsonUtility::isActionRequired(m_parsedJson, i_vpdFilePath,
+                                          "postAction", "collection"))
         {
-            throw std::runtime_error("Required post action failed for path " +
-                                     i_vpdFilePath +
-                                     " Aborting collection for this FRU");
+            if (!processPostAction(i_vpdFilePath, "collection", l_parsedVpd))
+            {
+                throw std::runtime_error(
+                    "Required post action failed for path " + i_vpdFilePath +
+                    " Aborting collection for this FRU");
+            }
         }
+
+        return l_parsedVpd;
     }
+    catch (std::exception& ex)
+    {
+        // If VPD parsing fails, and post fail action is required, execute it.
+        if (l_isPostFailActionRequired)
+        {
+            if (!jsonUtility::executePostFailAction(m_parsedJson, i_vpdFilePath,
+                                                    "collection"))
+            {
+                throw std::runtime_error("Post fail action failed for path " +
+                                         i_vpdFilePath +
+                                         " Aborting collection for this FRU");
```

Comment:
as we are not aborting anything, is the message should be ?
"VPD collection failed for the FRU " + i_vpdFilePath

Reply:
We end up in this `catch` block if
1. `vpdParser->parse()` throws when parsing fails.
2. `processPostAction()` throws if `parsedVpd` has no value.

Do you mean we should check for the type of exception caught and log accordingly here?
On a side note: I just noticed `processPostAction()` throws but it's doxygen doesn't mention `@throws`.

Reply:
ok, this is fine.

Reply:
@souvik1914581 In the case that you have added, in case "executePostFailAction" also fails then we loose the exception that was thrown by the parser path.
Could you please add that log as well to the log that you have added so that we don't miss the reason, why collection failed.

Reply:
Done